### DISCUSSION TITLE
Data format standard for frequency attribute

### DISCUSSION THREAD

**Discussion by vvujjini** [spdci/.github#25] [General]
Brining the discussion on this topic to github discussion to loop in larger audience. 

It is a good idea to explore "frequency" attribute data format between RRule and Crontab expressions. Both have good libraries and used widely. Crontab expression may be more straight forward due it it's popularity and existing support in solutions e.g., Kafka, Airflow, etc.,

@jeremi your thoughts between the two?

**Comment by jeremi** [comment]
For me, I prefer RRule for a few reasons:

- **Based on a standard**: iCalendar standard (RFC 5545)
- **Supports time zones**
- **Supports limited occurrences**
- **Broad set of libraries that support the format**
- **More advanced schedule support that could be useful for our use cases**:
  - **Last Specific Day of the Month**: Schedule an event on the last Friday of every month.
  - **N-th Specific Day of the Month**: Schedule events on the 3rd Wednesday of each month.
  - **Bi-weekly on a Specific Day**: Schedule an event every two weeks on Monday.
  - **Arbitrary Dates**: February 14, July 4, and December 25
  - **Quarterly on a Specific Day of the Month**: Schedule an event on the 15th of January, April, July, and October.

Implementations:
- **Python**: [dateutil](https://github.com/dateutil/dateutil)
- **Java**: [lib-recur](https://github.com/dmfs/lib-recur)
- **JS**: [rrule](https://github.com/jkbrzt/rrule)

**Reply by vvujjini** [reply]
Thanks @jeremi for calling out the features. Apart from iCal standard, all the other points you called out can be done using Crontab as well. :)

I'm okay with either and will go with the majority view. Definitely we should not attempt define this complex rules ourselves.

**Reply by jeremi** [reply]
@vvujjini As far as I know, the rules I listed are impossible with crontab without using associated scripts.
Cron does not support timezone and limited occurrence either.

**Reply by jeremi** [reply]
Crontab was designed for sysadmins, whereas RRule was designed for users to express the full complexity of scheduling. This is why RRule is used in all calendar implementations.
But yes if you edit the formula by hand, cron is slightly simpler. 
Another point: a cron formula can be converted to RRULE, the opposite is not always true.
